### ISSUE TITLE
Make rclone copy backup friendly

### ISSUE THREAD

**Issue by glensc** [rclone/rclone#5269]
#### What is your current rclone version (output from `rclone version`)?

```
rclone v1.55.0
- os/type: linux
- os/arch: arm
- go/version: go1.16.2
- go/linking: static
- go/tags: cmount
```

#### What problem are you are trying to solve?

`rclone copy`, when replacing files, seems to be overwriting them, i.e modify them in place rather creating new file (new inode). that is not much backup friendly where you make backups of hard linking previous directory:

```
cp -al workdir archive.YYYY-MM-DD
rclone copy gdrive: workdir
```

now if `gdrive/file1.jpg` was modified, then `archive.YYYY-MM-DD/file1.jpg` will be modified too because it's the same inode and `rclone copy` doesn't create new file.

#### How do you think rclone should be changed to solve that?

in other words, how `rclone copy` should operate, is to create new file and rename it, so the inode changes:
1. create temporary file `.target_name.XXXXX`
2. write new content to `.target_name.XXXXX`, verify stuff
3. rename `.target_name.XXXXX` to target_name

this is how `rsync` operates.

this should be done for reliability anyway, in case the process is interrupted before step 3, the original file will be still intact.

**Comment by ncw** [comment]
Yes I've noted this problem in the past. I think there might be an issue about it too...

Most rclone backends can't have partially uploaded files - a file is either uploaded successfully or not at all.

Notable exceptions are

- local backend
- sftp backend
- ftp backend
- some webdav backends

So what I'd propose is a new backend feature flag `CanHavePartialUploads`.

Then in `operations.Copy`

- if updating a file
- and the backend has `CanHavePartialUplaods`
- and the backend is capable of `Move` or `Copy`
- write the new file to a temporary file name
- on successful upload move the temp file over the existing file
- on failure remove the temp file

If rclone was not updating a file it would just upload it straight to the destination.

I'll note that rsync operates like this by default, but it has various flags to modify that behaviour like `--partial` `--partial-dir` `--inplace` which we might want to implement.

Does this seem sensible to you @glensc ?

Is this a feature you'd like to help implement?

**Comment by ncw** [comment]
There is lots of prior art in this ticket https://github.com/rclone/rclone/issues/3770

**Comment by glensc** [comment]
@ncw sorry, I don't have enough free time to learn Go here. I also have no idea about `rclone` internals. I trust you understood the problem I brought up.

**Comment by ncw** [comment]
This is fixed now. Any backends which partially uploaded files overwrite existing files will have there uploads done with a `.random-suffix.partial` file and then renamed into place. This can be overwritten with `--inplace`.

@eharris I think chunker takes care to avoid this by using random suffixes of its own but if not, please make another issue.

**Comment by eharris** [comment]
I'm guessing that this new suffix use affects filename length limits, including crypt.  Is that reflected in the docs somewhere?
The crypt docs at https://rclone.org/crypt/ suggest that the limit to avoid issues is 156 characters, but in my testing it is smaller than that.

**Comment by ncw** [comment]
> I'm guessing that this new suffix use affects filename length limits, including crypt.

If the file name is > 100 characters then the code which puts the prefix on won't make it any longer.

> Is that reflected in the docs somewhere?


> The crypt docs at https://rclone.org/crypt/ suggest that the limit to avoid issues is 156 characters, but in my testing it is smaller than that.

They do? I think <= 143 is the right number.

```
$ rclone test info --check-length secret:
// secret
maxFileLength = 143 // for 1 byte unicode characters
maxFileLength = 71 // for 2 byte unicode characters
maxFileLength = 47 // for 3 byte unicode characters
maxFileLength = 35 // for 4 byte unicode characters
```

**Comment by eharris** [comment]
That's a great command to know, I'm adding it to my notes!
I think 143 is also the right number for pure crypt overlay.  But it seems like most people do chunker+crypt, which with the default name_format is:
```
    maxFileLength = 126 // for 1 byte unicode characters
    maxFileLength = 63 // for 2 byte unicode characters
    maxFileLength = 42 // for 3 byte unicode characters
    maxFileLength = 31 // for 4 byte unicode characters
```
This is why I use a custom `name_format = *.rc=###` which gives:
```
    maxFileLength = 136 // for 1 byte unicode characters
    maxFileLength = 68 // for 2 byte unicode characters
    maxFileLength = 45 // for 3 byte unicode characters
    maxFileLength = 34 // for 4 byte unicode characters
```

**Comment by ncw** [comment]
I fixed the crypt docs - thanks for the heads-up.
